### ISSUE TITLE
AZ IoT C2D get lots of errors from C library when send messages via Django server

### ISSUE THREAD

**Issue by kwarodom** [Azure/azure-iot-sdk-python#223]
- **OS and version used:** <VERSION>  . Ubuntu 16.04

- **Python runtime used:** <VERSION>  Python2.7.12

- **SDK version used:** <VERSION> 

# Description of the issue:

I tried to develop my own backend using Django 1.11.16 with REST endpoint receiving POST method. This is to receive http POST request and then trigger a function with AZ C2D messaging to corresponding device ID. I can successfully send message to device with few requests. After that you will see my console log get exploding with lines and lines of errors from iot hub c library. Looking at my code i send http 204 response back to the request right after iot_messaging.send_sync is called. Is there anything i mistakenly did here? 

# Code sample exhibiting the issue:

<img width="835" alt="screen shot 2561-10-23 at 09 17 32" src="https://user-images.githubusercontent.com/2181490/47331516-6f98cc00-d6a6-11e8-8123-1f69265e3627.png">

# Console log of the issue:

<img width="899" alt="screen shot 2561-10-23 at 09 32 25" src="https://user-images.githubusercontent.com/2181490/47331604-cb635500-d6a6-11e8-8331-feee548b78e9.png">

**Comment by kwarodom** [comment]
Also when i tried to remove this 

   try:
            # Try Python 2.xx first
            raw_input("Press Enter to continue...\n")
        except:
            pass
            # Use Python 3.xx in the case of exception
            input("Press Enter to continue...\n")

before 
      -------
        iothub_messaging.close()

       --------

**Comment by az-iot-builder-01** [comment]
@kwarodom, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)

**Comment by pierreca** [comment]
it looks like you're closing the client before waiting for the asynchronous operation to complete. I don't know if that could cause these issues (I'll add the C SDK folks to this thread) but I'm pretty sure this is not what you should be doing. You should wait for the callback to be called for all send operations before you call close.

**Comment by jspaith** [comment]
@kwarodom - could you please make sure your call to `iothub_messaging.close` is called and let us know if that addresses the issue?  Looking at the code in question, it looks like in your for loop you're doing a 'return' after 1 pass (is that what you want by the way, since presumably you want to keep iterating?) so you may never execute the close call at end of function. 

@pierreca  - From code review of the C service SDK & Python boost, there's at least one problem.  If an app calls `IoTHubMessaging_Destroy` but not `IoTHubMessaging_Close` first, what's going to happen is that the _Destroy will take out the `Lock_Deinit(messagingClientInstance->LockHandle)`.  However, the `ScheduleWork_Thread` is still partying and when its lock is destroyed from underneath it - well, it's not happy.  

This points to a few issues, namely the Python boost layer should probably be calling ``IoTHubMessaging_Close` first in a destructor - if Close wasn't called yet - just as good programming practice.  But at same time the C SDK should have the `IoTHubMessaging_Destroy` itself wait for the thread to drain off if it's still active.  Opening a C github bug on that specifically in a moment...

So @kwarodom your code shouldn't be hitting this problem- underlying layers should handle cleanup better - but I think for a workaround for now just make sure you call the close and you should be OK.

**Comment by jspaith** [comment]
Created https://github.com/Azure/azure-iot-sdk-c/issues/693 to track.  And thanks @kwarodom for looking through your logs and calling this to our attention.

**Comment by kwarodom** [comment]
Here is what i found after trying to call iothub_messaging.close() still getting an error also, my device cannot receive not message as well.

<img width="1567" alt="screen shot 2561-10-26 at 15 17 44" src="https://user-images.githubusercontent.com/2181490/47554176-f06bf800-d932-11e8-9da4-8c05e5233bd8.png">

**Comment by jspaith** [comment]
@kwarodom - the send of the message happens asyncronously.  If you iothub_message.close() immediately after posting the message to send, then it's not going to have enough time to get the message out.

The code that you're thinking you're looking for is more along the lines of
for i in range(...): 
[tab]the logic you have up and including the iothub_messaging.send_async(...)
[tab]do not iothub_messaging.close()
[tab]do not "return response" -> you want the for loop to keep executing

After you've sent the messages you want, wait for the self.send_complete_callbacks to indicate that they've finished executing.  You should have a maximum amount of time you'll wait as a timeout just in case.

Only after the sends have been posted and received response from server (or you get a timeout in the failure case) should you be doing the `iothub_messaging.close()`.  You will (for now) want the explicit calls to `close` - at some point we'll make SDK's more forgiving where just destroying the iothub_messaging object will do the right thing whether on note `close` was invoked.

**Comment by neha-s-deshpande** [comment]
Hi, even after using iothub_messaging.close(), I am facing the same issue.

`Sending message: 0
Error: Time:Fri Aug 30 11:54:20 2019 File:/usr/sdk/src/c/uamqp/src/saslclientio.c Func:on_sasl_frame_received_callback Line:909 SASL handshake failed with code 02
Error: Time:Fri Aug 30 11:54:20 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:decode_ssl_received_bytes Line:797 SSL channel closed in decode_ssl_received_bytes.
Error: Time:Fri Aug 30 11:54:20 2019 File:/usr/sdk/src/c/uamqp/src/saslclientio.c Func:on_underlying_io_error Line:243 Error callback received in unexpected state
Error: Time:Fri Aug 30 11:54:20 2019 File:/usr/sdk/src/c/c-utility/adapters/tlsio_openssl.c Func:on_underlying_io_bytes_received Line:846 Error in decode_ssl_received_bytes.
Press Enter to continue...

send_complete_callback called with context : 0
messagingResult : OK
Error: Time:Fri Aug 30 11:54:41 2019 File:/usr/sdk/src/c/uamqp/src/saslclientio.c Func:saslclientio_send_async Line:1181 send called while not open
Error: Time:Fri Aug 30 11:54:41 2019 File:/usr/sdk/src/c/uamqp/src/connection.c Func:on_bytes_encoded Line:268 Cannot send encoded bytes
Error: Time:Fri Aug 30 11:54:41 2019 File:/usr/sdk/src/c/uamqp/src/saslclientio.c Func:saslclientio_close_async Line:1130 saslclientio_close called while not open
Error: Time:Fri Aug 30 11:54:41 2019 File:/usr/sdk/src/c/uamqp/src/connection.c Func:on_bytes_encoded Line:272 xio_close failed
Error: Time:Fri Aug 30 11:54:41 2019 File:/usr/sdk/src/c/uamqp/src/saslclientio.c Func:saslclientio_close_async Line:1130 saslclientio_close called while not open
Error: Time:Fri Aug 30 11:54:41 2019 File:/usr/sdk/src/c/uamqp/src/connection.c Func:connection_close Line:1437 xio_close failed
`

**Comment by jspaith** [comment]
@neha-s-deshpande - could you please open a new issue for this?  There are some SASL log errors from the underlying C implementation in the original issue, but they seem to be coming from a different section from the code from your log attached.

Please include the smallest possible repro of Python code along with which SDK version you're on, OS you're using, etc.

@pierreca - for this issue in particular, since original poster hasn't weighed in in quite a while I recommend we close this one out.

**Comment by neha-s-deshpande** [comment]
ok. I have opened issue @https://github.com/Azure/azure-iot-sdk-python/issues/291 .

**Comment by jebrando** [comment]
With the Python v2 reaching General Availability we are recommending users move away from the v1 SDK. 
We will be closing the v1 SDK issues whose features are covered by the v2 release. 
Feel free to open an new issue if you find an issue with the v2 SDK.
 
Thank you for your patronage

**Comment by az-iot-builder-01** [comment]
@jspaith, @kwarodom, @neha-s-deshpande, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
